bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,440 @@
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return true;
+      }
+
+      @Override
+      public Coder<Result> getDefaultOutputCoder() {
+        return new HBaseResultCoder();
+      }
+
+      @Override
+      public List<? extends BoundedSource<Result>> splitIntoBundles(long arg0, PipelineOptions arg1)
+          throws Exception {
+        return Collections.singletonList(this);
+      }
+
+      @Override
+      public void validate() {
+        checkNotNullOrEmpty(configuration.getProjectId(), "project");
+        checkNotNullOrEmpty(configuration.getZone(), "zone");
+        checkNotNullOrEmpty(configuration.getCluster(), "cluster");
+        checkNotNullOrEmpty(configuration.getTable(), "table");
+      }
+
+      @Override
+      public com.google.cloud.dataflow.sdk.io.BoundedSource.BoundedReader<Result> createReader(
+          PipelineOptions options) throws IOException {
+        return new CloudBigtableIO.Source.Reader(this, configuration, startKey, stopKey);
+      }
+    }
+
+    /**
+     * Reads rows for a specific {@link Table}, potentially filted by a Scan and/or a start/stop key
+     * range.
+     */
+    private static class Reader extends AbstractBoundedReader<Result> {
+      private final BoundedSource<Result> source;
+      private final Scan scan;
+      private final CloudBigtableScanConfiguration config;
+
+      private Connection connection;
+      private ResultScanner scanner;
+      private Table table;
+
+      private Result current;
+
+      public Reader(BoundedSource<Result> source, CloudBigtableScanConfiguration config,
+          byte[] startRow, byte[] endRow) throws IOException {
+        super();
+        this.source = source;
+        this.config = config;
+        if (startRow != null || endRow != null) {
+          this.scan = new Scan(config.getScan());
+          this.scan.setStartRow(startRow);
+          this.scan.setStopRow(endRow);
+        } else {
+          this.scan = config.getScan();
+        }
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        current = scanner.next();
+        return current != null;
+      }
+
+      @Override
+      public void close() throws IOException {
+        scanner.close();
+        table.close();
+        connection.close();
+      }
+
+      @Override
+      public Result getCurrent() throws NoSuchElementException {
+        return current;
+      }
+
+      @Override
+      public BoundedSource<Result> getCurrentSource() {
+        return source;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        connection = new BigtableConnection(config.toHBaseConfig());
+        table = connection.getTable(TableName.valueOf(config.getTable()));
+        scanner = table.getScanner(scan);
+        return advance();
```

Comment:
We shouldn't do an advance() call here. The sequence of operations for the user to get the first record will be start(), advance(), getCurrent(). If we do an advance() call in the start() method the user will miss the first record.